packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts

```diff
@@ -0,0 +1,228 @@
+
+import { assetDataUtils } from '../utils/asset_data_utils';
+
+export interface ForwarderSwapQuoteGetOutputOpts extends SwapQuoteGetOutputOpts {
+    feePercentage: number;
+    feeRecipient: string;
+    ethAmount?: BigNumber;
+}
+
+export interface ForwarderSwapQuoteExecutionOpts extends SwapQuoteExecutionOpts, ForwarderSwapQuoteGetOutputOpts {
+}
+
+const addAffiliateFeeToSwapQuoteInfo = (quoteInfo: SwapQuoteInfo, feePercentage: number): SwapQuoteInfo => {
+    const newQuoteInfo = _.clone(quoteInfo);
+    const affiliateFeeAmount = newQuoteInfo.takerTokenAmount.multipliedBy(feePercentage).integerValue(BigNumber.ROUND_CEIL);
+    const newFeeAmount = newQuoteInfo.feeTakerTokenAmount.plus(affiliateFeeAmount);
+    newQuoteInfo.feeTakerTokenAmount = newFeeAmount;
+    newQuoteInfo.totalTakerTokenAmount = newFeeAmount.plus(newQuoteInfo.takerTokenAmount);
+    return newQuoteInfo;
+};
+
+const addAffiliateFeeToSwapQuote = (quote: SwapQuote, feePercentage: number): SwapQuote => {
+    const newQuote = _.clone(quote);
+    newQuote.bestCaseQuoteInfo = addAffiliateFeeToSwapQuoteInfo(newQuote.bestCaseQuoteInfo, feePercentage);
+    newQuote.worstCaseQuoteInfo = addAffiliateFeeToSwapQuoteInfo(newQuote.worstCaseQuoteInfo, feePercentage);
+    return newQuote;
+};
+
+export class ForwarderSwapQuoteConsumer implements SwapQuoteConsumer {
+
+    public readonly provider: ZeroExProvider;
+    public readonly networkId: number;
+
+    private readonly _contractWrappers: ContractWrappers;
+
+    constructor(
+        supportedProvider: SupportedProvider,
+        options: Partial<SwapQuoteConsumerOpts> = {},
+    ) {
+        const { networkId } = _.merge(
+            {},
+            constants.DEFAULT_ASSET_SWAP_QUOTER_OPTS,
+            options,
+        );
+        const provider = providerUtils.standardizeOrThrow(supportedProvider);
+        assert.isNumber('networkId', networkId);
+        this.provider = provider;
+        this.networkId = networkId;
+        this._contractWrappers = new ContractWrappers(this.provider, {
+            networkId,
+        });
+    }
+
+    public getCalldataOrThrow = (quote: SwapQuote, opts: Partial<ForwarderSwapQuoteGetOutputOpts>): CalldataInformation => {
+        if (!this.isValidForwarderSwapQuote(quote)) {
+            throw new Error(AssetSwapQuoterError.InvalidForwarderSwapQuote);
+        }
+        const { params, to, value } = this.getSmartContractParamsOrThrow(quote, opts);
+        const methodAbi = utils.getMethodAbiFromContractAbi(this._contractWrappers.forwarder.abi, 'marketBuyOrdersWithEth') as MethodAbi;
+        const abiEncoder = new AbiEncoder.Method(methodAbi);
+        const args = [params.orders, params.makerAssetFillAmount, params.signatures, params.feeOrders, params.feeSignatures, params.feePercentage, params.feeRecipient];
+        const calldataHexString = abiEncoder.encode(args);
+        return {
+            calldataHexString,
+            to,
+            value,
+        };
+    }
+
```

Comment:
Nit: Remove these new lines. I believe this is an [unenforced coding convention](https://github.com/0xProject/0x-monorepo/blob/development/CONTRIBUTING.md#unenforced-coding-conventions). You should also set up the [auto-formatter](https://github.com/0xProject/0x-monorepo/blob/development/CONTRIBUTING.md#development-tooling) though.